numpy/core/src/multiarray/number.c

```diff
@@ -930,22 +932,22 @@ array_index(PyArrayObject *v)
 
 
 NPY_NO_EXPORT PyNumberMethods array_as_number = {
-    .nb_add = (binaryfunc)array_add,
-    .nb_subtract = (binaryfunc)array_subtract,
-    .nb_multiply = (binaryfunc)array_multiply,
-    .nb_remainder = (binaryfunc)array_remainder,
-    .nb_divmod = (binaryfunc)array_divmod,
+    .nb_add = array_add,
+    .nb_subtract = array_subtract,
+    .nb_multiply = array_multiply,
+    .nb_remainder = array_remainder,
+    .nb_divmod = array_divmod,
     .nb_power = (ternaryfunc)array_power,
     .nb_negative = (unaryfunc)array_negative,
     .nb_positive = (unaryfunc)array_positive,
     .nb_absolute = (unaryfunc)array_absolute,
     .nb_bool = (inquiry)_array_nonzero,
     .nb_invert = (unaryfunc)array_invert,
-    .nb_lshift = (binaryfunc)array_left_shift,
-    .nb_rshift = (binaryfunc)array_right_shift,
-    .nb_and = (binaryfunc)array_bitwise_and,
-    .nb_xor = (binaryfunc)array_bitwise_xor,
-    .nb_or = (binaryfunc)array_bitwise_or,
+    .nb_lshift = array_left_shift,
+    .nb_rshift = array_right_shift,
+    .nb_and = array_bitwise_and,
+    .nb_xor = array_bitwise_xor,
+    .nb_or = array_bitwise_or,
```

Comment:
Was removing the cast here the motivation for this PR?

Reply:
This was just a side effect.   The reason for this was that I got a bit confused by it when I was trying how to add special handling for Python scalars (for the discussion about making `np.array(1, dtype=uint8) + 100` return `uint8` no matter the value).
(Since there you have to check for both the first and second argument being a Python scalar)